fix for mask generation from >1 channel

smguo opened the pull request:
This PR is intended to fix the summing dimension of image stack when >1 channel are used to generate masks. Each channel is thresholded first and then summed so the masks are not biased by the bright channel. The fix has been tested for 2D images but not 3D

jennyfolkesson commented:
@smguo I like the option of saving masks as uint8 pngs instead of npys so you can open them with any viewer. I think it can be an either/or option though instead of having duplicate images with different formats, as we can easily read png masks as well as npys in preprocessing.

FYI we've added CI tests, you can check Travis details above to see how your build went.
It looks like /tests/utils/mp_utils_tests.py tests are failing, I'll review when you've fixed that!

smguo commented:
Thanks @jennyfolkesson! Yes, it makes more sense to save masks as either uint8 pngs or npys rather than both. I have changed that and also updated test functions so they don't fail

jennyfolkesson commented on micro_dl/utils/mp_utils.py:
```diff
@@ -47,53 +48,54 @@ def create_save_mask(input_fnames,
+    elif mask_ext == 'png':
+        file_name = file_name[:-3] + 'png'
+        # Covert mask to uint8
+        mask = mask.astype(np.uint8) * (2 ** 8 - 1)
```
I think it's fine to just write 255. Or if you want to be fancy you can use np.iinfo

anithapk replied:
Wondering how this will be read with openCV, will the unique values be 0 and 255. That might an issue here: https://github.com/czbiohub/microDL/blob/876085b13060650ce955a202cf09506f120974ad/micro_dl/utils/tile_utils.py#L179 as the mean is not a volume fraction or percent anymore!

jennyfolkesson replied:
Good point. I think it should be fine if the images are converted to bool after reading them, e.g. here: https://github.com/czbiohub/microDL/blob/876085b13060650ce955a202cf09506f120974ad/micro_dl/utils/tile_utils.py#L29

jennyfolkesson commented on micro_dl/utils/mp_utils.py:
```diff
@@ -47,53 +48,54 @@ def create_save_mask(input_fnames,
-                                               str_elem_radius)
+    )
+    masks = []
+    for idx in range(im_stack.shape[-1]):
```
I guess this changes the behavior of masking? Instead of adding images together then masking the summed image you add the generated masks together?
At the very least I think the multi-channel behavior needs to be clarified in the docstring.

smguo replied:
Yes, summing images first doesn't work well when the intensities of channels are not on similar scale. I documented the behavior in the docstring
